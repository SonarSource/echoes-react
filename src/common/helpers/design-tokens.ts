/*
 * Echoes React
 * Copyright (C) 2023-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import designTokensColors from '~generated/design-tokens-colors.json';

export type DesignTokens = keyof typeof designTokensColors;

export type DesignTokensColors = {
  [K in DesignTokens]: K extends `echoes-color-${string}` ? K : never;
}[DesignTokens];
