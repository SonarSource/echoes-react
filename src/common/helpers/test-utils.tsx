/*
 * Echoes React
 * Copyright (C) 2023-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { RenderOptions, render as rtlRender } from '@testing-library/react';
import userEvent, { Options as UserEventsOptions } from '@testing-library/user-event';
import React, { ComponentProps, PropsWithChildren } from 'react';
import { IntlProvider } from 'react-intl';
import { PropsWithLabels } from '~types/utils';
import { TooltipProvider } from '../../components/tooltip';

export function render(
  ui: React.ReactElement,
  options?: RenderOptions,
  userEventOptions?: UserEventsOptions,
) {
  return {
    ...rtlRender(ui, { wrapper: ContextWrapper, ...options }),
    user: userEvent.setup(userEventOptions),
  };
}

function ContextWrapper({ children }: PropsWithChildren<{}>) {
  return (
    <IntlProvider defaultLocale="en-us" locale="en-us">
      <TooltipProvider delayDuration={0}>{children}</TooltipProvider>
    </IntlProvider>
  );
}

export type OmitPropsWithLabels<T extends React.JSXElementConstructor<any>> = Partial<
  Omit<ComponentProps<T>, keyof PropsWithLabels<{}>>
> &
  PropsWithLabels<{}>;
