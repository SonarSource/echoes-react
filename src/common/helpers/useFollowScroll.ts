/*
 * Echoes React
 * Copyright (C) 2023-2025 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { useEffect, useState } from 'react';
import { throttle, THROTTLE_SHORT_DELAY } from './utils';

/**
 * A React hook that tracks the scroll position of the document.
 * It returns an object containing the current scroll position (left and top).
 * The scroll position is updated at a throttled rate to improve performance.
 *
 * @returns An object with `left` and `top` properties representing the current scroll position.
 */
export function useFollowScroll(enabled = true) {
  const [left, setLeft] = useState(0);
  const [top, setTop] = useState(0);

  useEffect(() => {
    const followScroll = throttle(() => {
      if (document.documentElement) {
        setLeft(document.documentElement.scrollLeft);
        setTop(document.documentElement.scrollTop);
      }
    }, THROTTLE_SHORT_DELAY);

    if (enabled) {
      document.addEventListener('scroll', followScroll);
    }

    return () => document.removeEventListener('scroll', followScroll);
  }, [enabled]);

  return { left, top };
}
