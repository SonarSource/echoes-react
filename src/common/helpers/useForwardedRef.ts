/*
 * Echoes React
 * Copyright (C) 2023-2025 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { type ForwardedRef, useState } from 'react';

/**
 * This hook may be used to intercept a forwarded ref, providing a local ref
 * that can be used in the component.
 *
 * **Example**
 *
 * ```typescript
 * const Input = forwardRef<HTMLInputElement>((props, forwardedRef) => {
 *   const [ref, setRef] = useForwardedRef(forwardedRef);
 *   return <input ref={setRef} />
 * });
 * ```
 */
export function useForwardedRef<T>(forwardedRef: ForwardedRef<T>) {
  const [ref, setRef] = useState<T | null>(null);

  const setForwardedRef = (element: T | null) => {
    if (typeof forwardedRef === 'function') {
      forwardedRef(element);
    } else if (forwardedRef) {
      forwardedRef.current = element;
    }

    setRef(element);
  };

  return [ref, setForwardedRef] as const;
}
