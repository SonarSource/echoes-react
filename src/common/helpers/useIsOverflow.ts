/*
 * Echoes React
 * Copyright (C) 2023-2025 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { DependencyList, MutableRefObject, useLayoutEffect, useState } from 'react';

/**
 * This hook checks if the content of a given ref is overflowing its container horizontally.
 * It returns a boolean indicating whether the content is overflowing.
 *
 * @param forwardedRef - The ref to check for overflow.
 * @param deps - Optional dependencies array to re-compute the overflow state.
 */
export function useIsOverflow(
  forwardedRef: MutableRefObject<HTMLElement | null>,
  deps: DependencyList = [],
) {
  const [isOverflow, setIsOverflow] = useState<boolean | undefined>(undefined);

  useLayoutEffect(() => {
    if (forwardedRef.current) {
      setIsOverflow(forwardedRef.current.scrollWidth > forwardedRef.current.offsetWidth);
    }
    // We want to be able to pass an array of dependencies to re-compute the overflow state
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [forwardedRef, ...deps]);

  return [isOverflow] as const;
}
