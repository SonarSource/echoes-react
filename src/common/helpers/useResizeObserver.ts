/*
 * Echoes React
 * Copyright (C) 2023-2025 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { RefObject, useCallback, useEffect, useState } from 'react';

export function useResizeObserver(ref: RefObject<HTMLElement>) {
  const [width, setWidth] = useState<number>();
  const [height, setHeight] = useState<number>();

  const handleResize = useCallback<ResizeObserverCallback>((entries) => {
    if (!Array.isArray(entries)) {
      return;
    }

    const entry = entries[0];
    setWidth(entry.contentRect.width);
    setHeight(entry.contentRect.height);
  }, []);

  useEffect(() => {
    if (!ref.current) {
      return undefined;
    }

    const RO = new ResizeObserver(handleResize);
    RO.observe(ref.current);

    return () => {
      RO.disconnect();
    };
  }, [handleResize, ref]);

  return { width, height };
}
