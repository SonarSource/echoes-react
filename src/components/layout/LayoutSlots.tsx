/*
 * Echoes React
 * Copyright (C) 2023-2025 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import styled from '@emotion/styled';
import { CSSProperties, forwardRef, PropsWithChildren } from 'react';
import { cssVar } from '~utils/design-tokens';
import { AsideSize, ContentGridArea, GlobalGridArea, PageGridArea, PageWidth } from './LayoutTypes';

const AsideSizeStyles: Record<AsideSize, CSSProperties> = {
  [AsideSize.small]: { width: cssVar('layout-aside-width-small') },
  [AsideSize.medium]: { width: cssVar('layout-aside-width-medium') },
  [AsideSize.large]: { width: cssVar('layout-aside-width-large') },
};

/*
 * Layer 1 Grid components
 */
export const BannerContainer = styled.div`
  grid-area: ${GlobalGridArea.banner};

  z-index: 1; // Ensure the banners are showing over the content
`;
BannerContainer.displayName = 'BannerContainer';

export const ContentGrid = styled.div`
  grid-area: ${GlobalGridArea.content};

  position: relative;
  overflow-y: hidden;
  isolation: isolate; // Reset stacking context

  display: grid;
  grid-template-columns: auto 1fr;
  grid-template-rows: auto 1fr;
  grid-template-areas:
    '${ContentGridArea.header} ${ContentGridArea.header}'
    '${ContentGridArea.aside} ${ContentGridArea.page}';

  padding-left: ${cssVar('dimension-space-300')};
`;
ContentGrid.displayName = 'ContentGrid';

/*
 * Layer 2 Grid components
 * (ContentGrid)
 */
export interface AsideProps {
  className?: string;
  size: `${AsideSize}`;
}
export const AsideLeft = forwardRef<HTMLDivElement, PropsWithChildren<AsideProps>>((props, ref) => {
  const { children, size, ...restProps } = props;
  return (
    <StyledAside {...restProps} ref={ref} style={AsideSizeStyles[size]}>
      {children}
    </StyledAside>
  );
});
AsideLeft.displayName = 'AsideLeft';

const StyledAside = styled.div`
  grid-area: ${ContentGridArea.aside};
  overflow-y: auto;

  box-sizing: border-box;
  padding-left: ${cssVar('dimension-space-50')};
  padding-right: ${cssVar('dimension-space-100')};
  padding-top: ${cssVar('dimension-space-300')};
  margin-right: ${cssVar('dimension-space-200')};
`;
StyledAside.displayName = 'StyledAside';

export interface PageGridProps {
  className?: string;
  width: `${PageWidth}`;
}

export const PageGrid = forwardRef<HTMLDivElement, PropsWithChildren<PageGridProps>>(
  (props, ref) => {
    const { children, width, ...restProps } = props;
    return (
      <PageGridContainer {...restProps} ref={ref}>
        <PageGridInner style={PAGE_WIDTH_STYLES[width]}>{children}</PageGridInner>
      </PageGridContainer>
    );
  },
);
PageGrid.displayName = 'PageGrid';

const PageGridContainer = styled.div`
  grid-area: ${ContentGridArea.page};
  overflow-y: auto;

  display: grid;
  grid-template-rows: auto 1fr auto;
  grid-template-areas:
    '${PageGridArea.header}'
    '${PageGridArea.main}'
    '${PageGridArea.footer}';
`;
PageGridContainer.displayName = 'PageGridContainer';

const PageGridInner = styled.div`
  display: grid;
  grid-template-rows: auto 1fr auto;
  grid-template-areas:
    '${PageGridArea.header}'
    '${PageGridArea.main}'
    '${PageGridArea.footer}';
`;
PageGridInner.displayName = 'PageGridInner';

const PAGE_WIDTH_STYLES: Record<PageWidth, CSSProperties> = {
  [PageWidth.default]: {
    marginLeft: 'auto',
    marginRight: 'auto',
    maxWidth: cssVar('layout-sizes-max-width-default'),
  },

  [PageWidth.fluid]: {},
};

export const PageContent = styled.div`
  grid-area: ${PageGridArea.main};

  padding-top: ${cssVar('dimension-space-300')};
  padding-bottom: ${cssVar('dimension-space-300')};
  padding-right: ${cssVar('dimension-space-300')};
`;
PageContent.displayName = 'PageContent';

export const PageFooter = styled.div`
  grid-area: ${PageGridArea.footer};

  display: flex;
  justify-content: space-between;
  align-items: center;

  padding-top: ${cssVar('dimension-space-200')};
  padding-bottom: ${cssVar('dimension-space-300')};
  padding-right: ${cssVar('dimension-space-300')};
`;
PageFooter.displayName = 'PageFooter';
