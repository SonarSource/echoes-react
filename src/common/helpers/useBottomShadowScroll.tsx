/*
 * Echoes React
 * Copyright (C) 2023-2025 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import styled from '@emotion/styled';
import { RefObject, useCallback, useEffect, useState } from 'react';
import { cssVar } from '~utils/design-tokens';
import { useResizeObserver } from './useResizeObserver';

/**
 * Custom hook to manage the visibility of a bottom shadow on a scrollable container.
 * The shadow indicates that there is more content to scroll to.
 * @param scrollableContainerRef - Ref to the scrollable container element, its height and scroll event are monitored to recompute the shadow
 * @param resizableContentRef - Ref to the resizable content element inside the scrollable container, its height is monitored to recompute the shadow
 * @returns {boolean} Whether the bottom shadow should be shown
 */
export function useBottomShadowScroll(
  scrollableContainerRef: RefObject<HTMLElement>,
  resizableContentRef: RefObject<HTMLElement>,
) {
  const [showBottomShadow, setShowBottomShadow] = useState(false);

  // Monitor the height of the scrollable container, and recompute the shadow if it changes
  const { height: scrollableContainerHeight } = useResizeObserver(scrollableContainerRef);
  // Monitor the height of the resizable content, and recompute the shadow if it changes, because it may move the scroll position
  const { height: resizableContentHeight } = useResizeObserver(resizableContentRef);

  const updateBottomShadowScroll = useCallback(() => {
    if (scrollableContainerRef.current) {
      const { scrollHeight, clientHeight, scrollTop } = scrollableContainerRef.current;
      setShowBottomShadow(scrollTop + clientHeight < scrollHeight - 8); // -8px for bottom padding
    }
  }, [scrollableContainerRef]);

  // Compute the bottom shadow visibility on first render and attach the scroll event listener
  useEffect(() => {
    const scrollableContainerRefNode = scrollableContainerRef.current;
    if (!scrollableContainerRefNode) {
      return undefined;
    }

    // Re-compute the bottom shadow visibility on scroll event
    scrollableContainerRefNode.addEventListener('scroll', updateBottomShadowScroll);

    // Initial computation
    updateBottomShadowScroll();

    return () => {
      scrollableContainerRefNode.removeEventListener('scroll', updateBottomShadowScroll);
    };
  }, [scrollableContainerRef, updateBottomShadowScroll]);

  // Re-compute the bottom shadow visibility on container and content resize
  useEffect(updateBottomShadowScroll, [
    resizableContentHeight,
    scrollableContainerHeight,
    updateBottomShadowScroll,
  ]);

  return [showBottomShadow] as const;
}

export const BottomShadowScroll = styled.div`
  position: absolute;
  bottom: 0;
  left: 0;
  right: 0;

  min-height: ${cssVar('dimension-height-400')};

  border-bottom: ${cssVar('border-width-default')} solid ${cssVar('color-border-weak')};
  background: radial-gradient(
      farthest-side at 50% 100%,
      ${cssVar('bottom-scroll-colors-shadow-gradient')},
      ${cssVar('color-support-transparent')}
    )
    center bottom;
`;
BottomShadowScroll.displayName = 'BottomShadowScroll';
