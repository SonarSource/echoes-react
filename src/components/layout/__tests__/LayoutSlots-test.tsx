/*
 * Echoes React
 * Copyright (C) 2023-2025 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { screen } from '@testing-library/react';
import { render } from '../../../common/helpers/test-utils';
import { AsideLeft, AsideSize, ContentGrid, ContentWidth } from '../LayoutSlots';

describe('ContentGrid', () => {
  it.each([
    [ContentWidth.fixed, 'var(--echoes-layout-sizes-max-width-default)'],
    [ContentWidth.legacy, 'var(--echoes-layout-sizes-max-width-large)'],
  ])('should render correctly when %s', (width, expected) => {
    render(<ContentGrid width={width}>content</ContentGrid>);

    expect(screen.getByText('content')).toHaveStyle({ maxWidth: expected });
  });

  it('should render correctly when fluid', () => {
    render(<ContentGrid width={ContentWidth.fluid}>content</ContentGrid>);

    expect(screen.getByText('content')).not.toHaveStyle({
      maxWidth: 'var(--echoes-layout-sizes-max-width-default)',
    });
  });
});

describe('AsideLeft', () => {
  it.each([
    [AsideSize.small, 'var(--echoes-layout-aside-width-small)'],
    // Useless to run these (see below):

    // [AsideSize.medium, 'var(--echoes-layout-aside-width-medium)'],
    // [AsideSize.large, 'var(--echoes-layout-aside-width-large)'],
  ])('should render correctly when %s', (size, _expected) => {
    render(<AsideLeft size={size}>content</AsideLeft>);

    /* There is a bug with emotion-jest preventing the styles from being included
     * This means we can't validate the style.
     * https://github.com/emotion-js/emotion/issues/3178
     *
     * This is what we'd like to check:
     * expect(screen.getByText('content')).toHaveStyle({ width: expected });
     *
     * Instead, we do this empty check...
     */
    expect(screen.getByText('content')).toMatchSnapshot();
  });
});
