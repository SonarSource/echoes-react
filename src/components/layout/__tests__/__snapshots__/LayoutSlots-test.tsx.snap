// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`AsideLeft should render correctly when small 1`] = `
.emotion-0 {
  grid-area: aside;
  overflow-y: auto;
}

<div
  class="emotion-0 emotion-1"
>
  content
</div>
`;
